### PR TITLE
[SPARK-26761][SQL][R] Vectorized R gapply() implementation

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR targets to add vectorized `gapply()` in R, Arrow optimization.

This can be tested as below:

```bash
$ ./bin/sparkR --conf spark.sql.execution.arrow.enabled=true
```

```r
df <- createDataFrame(mtcars)
collect(gapply(df,
               "gear",
               function(key, group) {
                 data.frame(gear = key[[1]], disp = mean(group$disp) > group$disp)
               },
               structType("gear double, disp boolean")))
```

### Requirements
  - R 3.5.x 
  - Arrow package 0.12+
    ```bash
    Rscript -e 'remotes::install_github("apache/arrow@apache-arrow-0.12.0", subdir = "r")'
    ```

**Note:** currently, Arrow R package is not in CRAN. Please take a look at ARROW-3204.
**Note:** currently, Arrow R package seems not supporting Windows. Please take a look at ARROW-3204.

### Benchmarks

**Shall**

```bash
sync && sudo purge
./bin/sparkR --conf spark.sql.execution.arrow.enabled=false --driver-memory 4g
```

```bash
sync && sudo purge
./bin/sparkR --conf spark.sql.execution.arrow.enabled=true --driver-memory 4g
```

**R code**

```r
rdf <- read.csv("500000.csv")
df <- cache(createDataFrame(rdf))
count(df)


test <- function() {
  options(digits.secs = 6) # milliseconds
  start.time <- Sys.time()
  count(cache(gapply(df,
                     "Month_of_Joining",
                     function(key, group) { group },
                     schema(df))))
  end.time <- Sys.time()
  time.taken <- end.time - start.time
  print(time.taken)
}


test()
```

**Data (350 MB):**

```r
object.size(read.csv("500000.csv"))
350379504 bytes
```

"500000 Records"  http://eforexcel.com/wp/downloads-16-sample-csv-files-data-sets-for-testing/

**Results**

```
Time difference of 5.195142 mins
```

```
Time difference of 12.64185 secs
```

The performance improvement was around **2465%**.

### Limitations

- For now, Arrow optimization with R does not support when the data is `raw`, and when user explicitly gives float type in the schema. They produce corrupt values.

- Due to ARROW-4512, it cannot send and receive batch by batch. It has to send all batches in Arrow stream format at once. It needs improvement later.

## How was this patch tested?

Unit tests were added

**TODOs:**
- [x] Draft codes
- [x] make the tests passed
- [x] make the CRAN check pass
- [x] Performance measurement
- [x] Supportability investigation (for instance types)

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
I keep forgetting which branch I was working on ... :( . Let me leave the PR open and work here.

Let me leave cc @BryanCutler, @viirya, @felixcheung anyway.

**Comment by HyukjinKwon** [comment]
Hey guys, I am sure it's ready. Adding more people @rxin, @gatorsmile, @shivaram, @falaki for visibility.

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: Codes here are almost not changed. I just exposed some method to inherit it in ArrowRRunner. This code needs refactoring as a common parent like PythonRunner but I don't target it here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I am going to fix the doc later when this R with Arrow optimization is finished.

**Comment by HyukjinKwon** [comment]
I'll start `dapply` once this one gets merged. I think codes can be shared.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: This protocol is different with Pandas's grouped map UDF one. Pandas grouped UDF sends arguments position to find the grouping keys but here I just decided to follow existing protocol in `gapply()`. Keys are relatively small so it won't affect the performance a lot anyway.

File: R/pkg/inst/worker/worker.R
Comment (HyukjinKwon) [inline R/pkg/inst/worker/worker.R]: Actually, according to the link above, the fastest way is convert them as matrix. I tried it before but looks it's not able to keep its types apparently. I tried other ways too but this seems to be the only way.

File: R/pkg/R/deserialize.R
Comment (HyukjinKwon) [inline R/pkg/R/deserialize.R]: Arrow seems to only provide an API that converts Arrow batch to Tibble (not `data.frame`). They are similar but different. I converted it to `as.data.frame` to be consistent with existing `gapply`. The cost looks very cheap since some doc say `as.data.frame` works like a thin wrapper in this case.

File: R/pkg/tests/fulltests/test_sparkSQL.R
Comment (HyukjinKwon) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: I am willing to make a wrapper for this `spark.sql.execution.arrow.enabled` set/unset but I would like to target this separately later if you guys won't mind.

**Comment by HyukjinKwon** [comment]
All tests passed except CRAN. I'm gonna take a look and fix soon.

File: R/pkg/tests/fulltests/test_sparkSQL.R
Comment (felixcheung) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: If this is mostly in tests I wouldn't bother with another method...

File: R/pkg/R/deserialize.R
Comment (felixcheung) [inline R/pkg/R/deserialize.R]: these are pretty nasty... maybe put all these into a R file/method

```
  requireNamespace1 <- requireNamespace
  requireNamespace1("arrow", quietly = TRUE)
  RecordBatchStreamReader <- get(
    "RecordBatchStreamReader", envir = asNamespace("arrow"), inherits = FALSE)
  as_tibble <- get("as_tibble", envir = asNamespace("arrow"))
```

File: R/pkg/R/group.R
Comment (felixcheung) [inline R/pkg/R/group.R]: use `sparkR.conf(key, defaultValue)` instead? otherwise it will call `stop()` if the config is not set?

File: R/pkg/R/group.R
Comment (felixcheung) [inline R/pkg/R/group.R]: IMO it will be more clear to say `gapply/gapplyCollect`

File: R/pkg/inst/worker/worker.R
Comment (felixcheung) [reply R/pkg/inst/worker/worker.R]: shouldn't `rbind(outputs)` work here?

**Review by felixcheung — COMMENTED** [review]
cool!

**Comment by HyukjinKwon** [comment]
retest this please

File: R/pkg/inst/worker/worker.R
Comment (felixcheung) [reply R/pkg/inst/worker/worker.R]: and be aware of string columns https://github.com/apache/spark/blob/master/R/pkg/inst/worker/worker.R#L35

File: R/pkg/R/deserialize.R
Comment (HyukjinKwon) [reply R/pkg/R/deserialize.R]: Yes .. I actually tried to pull those out.. but I ended up with leave them as are .. I target to remove those weird hacks as soon as Arrow is on CRAN which I estimate before Spark 3.0. I promise I clean up those before Spark 3.0.

File: R/pkg/R/group.R
Comment (HyukjinKwon) [reply R/pkg/R/group.R]: Actually, for this specifically, unlike other configurations, `sparkR.conf` is guaranteed to return a default value for the configuration that exists in `SQLConf` if I didn't misread. It should be okay to leave the default value in `SQLConf`.

File: R/pkg/inst/worker/worker.R
Comment (HyukjinKwon) [reply R/pkg/inst/worker/worker.R]: Seems `rbind(iris(dataframe, ...)` not working .. :(. 

Yea, in terms of strings as factors ..  for input, it should be okay about `stringsAsFactors` because it's directly converted from Arrow batch.

For output, yes, I think there might be a concern about strings as factors. I currently assume the given function should always return the same `data.frame`s to combine. So I guess it's okay. I would like to leave this concern separate if you don't mind. We don't take care of factors in JVM side anyway.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: R/pkg/inst/worker/worker.R
Comment (felixcheung) [inline R/pkg/inst/worker/worker.R]: append should be faster
```
> outputs <- list()
> output <- 1
> outputs <- append(outputs, output)
> outputs
[[1]]
[1] 1

> outputs <- append(outputs, output)
> outputs
[[1]]
[1] 1

[[2]]
[1] 1
```

File: R/pkg/inst/worker/worker.R
Comment (felixcheung) [reply R/pkg/inst/worker/worker.R]: shouldn't we turn it off (as `stringsAsFactors = FALSE`)? 

If I recall, we don't handle factor to JVM correctly, and compute() from the user function doesn't **really** have to return a data.frame per se.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala
Comment (felixcheung) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala]: doesnt need `CatalystSerde.serialize ... RowEncoder` here?

**Comment by HyukjinKwon** [comment]
All the test failures look unrelated.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala
Comment (felixcheung) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: natches?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala
Comment (felixcheung) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: this and hasNext is a bit weird, but I assume there's a reason for it..

File: R/pkg/R/group.R
Comment (HyukjinKwon) [inline R/pkg/R/group.R]: @felixcheung, I was double checking one by one and realised that I need some more fixes for `gapplyCollect()`. Currently, I disabled it when Arrow is enabled:

```r
> df <- createDataFrame(mtcars)
> gapplyCollect(df,
+               "gear",
+               function(key, group) {
+                 data.frame(gear = key[[1]], disp = mean(group$disp) > group$disp)
+               })
Error in gapplyInternal(x, func, NULL) :
  Arrow optimization does not support gapplyCollect yet. Please use 'collect' and 'gapply' APIs instead.
```

 I need few line changes (I guess between 10 ~ 20 lines) to support gapply but let me do this separately with a separate set of tests.

I file a JIRA here, https://issues.apache.org/jira/browse/SPARK-26858. I will do this too as soon as this PR gets merged.

**Comment by felixcheung** [comment]
what about https://github.com/apache/spark/commit/e8982ca7ad94e98d907babf2d6f1068b7cd064c6#diff-b11442485f6b77bf47b58b4747321638R267 can this be changed from file to stream also?

**Comment by HyukjinKwon** [comment]
> what about [e8982ca#diff-b11442485f6b77bf47b58b4747321638R267](https://github.com/apache/spark/commit/e8982ca7ad94e98d907babf2d6f1068b7cd064c6#diff-b11442485f6b77bf47b58b4747321638R267) can this be changed from file to stream also?

Yes.. it actually used file on purpose actually.... if I use a file to transfer the data (instead of socket), I can send batch by batch in a streaming matter but if I use socket, I should buffer all the batches and then send them at once due to the Arrow R API limitation (ARROW-4512).

I used file there because the existing protocol at R DataFrame -> Spark DataFrame was already using a file .. but looks all other protocol are using sockets. So, I had to use socket here to match the protocol although I have to buffer all the batches due to Arrow R API limitation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala]: D'oh.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala]: Yea, that's actually related with `gapplyCollect`. I have a neat idea to avoid unnecessary de/encode. I will make a fix after this PR, when I fix `gapplyCollect` against https://issues.apache.org/jira/browse/SPARK-26858.

File: R/pkg/inst/worker/worker.R
Comment (HyukjinKwon) [reply R/pkg/inst/worker/worker.R]: Yup, turning off should be okay too. Let me turn off then.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala]: I agree the codes here looks a bit messy .. I mimicked the code structure from `PythonRunner`s for a possibility of deduplicating later. Should be okay.

File: R/pkg/inst/worker/worker.R
Comment (HyukjinKwon) [reply R/pkg/inst/worker/worker.R]:
```r
outputA <- list()
outputB <- list()

testA <- function() {
  options(digits.secs = 6) # milliseconds
  start.time <- Sys.time()
  output <- list()
  for (i in 1:10000) {
    output[[length(output) + 1L]] <- 1
  }
  end.time <- Sys.time()
  time.taken <- end.time - start.time
  print(time.taken)
  output
}



testB <- function() {
  options(digits.secs = 6) # milliseconds
  start.time <- Sys.time()
  output <- list()
  for (i in 1:10000) {
    output <- append(output, 1)
  }
  end.time <- Sys.time()
  time.taken <- end.time - start.time
  print(time.taken)
  output
}

require(testthat)

outputA <- testA()
outputB <- testB()

expect_equal(outputA, outputB)
```

Looks direct assignement is faster:

```
Time difference of 0.008903027 secs
Time difference of 0.5212681 secs
```

File: R/pkg/inst/worker/worker.R
Comment (HyukjinKwon) [reply R/pkg/inst/worker/worker.R]: Actually, setting `stringsAsFactors` to `FALSE` downgrades its performance roughly twice ( from 4 secs to 8 secs for the benchmark in the PR description). Since it's being ignored in JVM, shouldn't be a matter to leave it.

File: R/pkg/R/deserialize.R
Comment (felixcheung) [inline R/pkg/R/deserialize.R]: another approach
```
  if (!requireNamespace1("arrow", quietly = TRUE)) {
    stop("'arrow' package should be installed.")
  }
```

File: R/pkg/inst/worker/worker.R
Comment (felixcheung) [reply R/pkg/inst/worker/worker.R]: haha, ok ok ;)

**Comment by felixcheung** [comment]
actually I think file is "problematic" if you remember some problem from the last few months. if we can do streaming we should do that across R & python too, for data.frame or toPandas @BryanCutler FYI

**Review by felixcheung — APPROVED** [review]
LGTM-  I'm still worry that `stringsAsFactors` can be a problem as per comment, but I'll let you check.

File: R/pkg/R/deserialize.R
Comment (HyukjinKwon) [reply R/pkg/R/deserialize.R]: Let me fix it when I clean up this logic to be same and minimised.

**Comment by HyukjinKwon** [comment]
Thanks, Felix. I'm going to work on related changes like dapply, gapplyCollect, dapplyCollect as well. I will double check it for `stringsAsFactors`.

I am merging this to unblock other vectorized API works. Considering Arrow API is not installed in Jenkins, and the current tests all passes, I assume it doesn't harm to existing use cases.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
BTW, I am also going to document that this arrow optimization should be unstable, experimental and technical preview, and It's discouraged to be used in production for now later when I finish this jobs.

**Comment by HyukjinKwon** [comment]
I updated benchmark with bigger data that shows perf diff more clearly.

**Pull request closed without merging**
